### PR TITLE
rewrite parser to use syn 0.15

### PR DESCRIPTION
**Pull request by jmhain**
Saw the call for participation on TWIR for issue #17, so decided to give it a shot. Passes tests, but seems to be slower than the current syn 0.14-based parser.

### CODE REVIEW

**Comment by lukaslueg** [comment]
Thanks for the PR!

The following macro should produce two literals `'` and `a`, yet the new parser seems to skip the `'` and only produces one literal `a`. Could you look into this?

```rust
macro_rules! a { ('a) => { ... }; }
```

**Comment by jmhain** [comment]
> The following macro should produce two literals ' and a, yet the new parser seems to skip the ' and only produces one literal a. Could you look into this?

Yeah, it looks like `syn` stopped parsing `'` as a `Punct`, and instead only as part of a lifetime. Added a commit with a workaround.

File: src/parser.rs [hunk 1]
```diff
@@ -74,109 +76,227 @@ pub enum Fragment {
 }
 
 macro_rules! delimited {
-    ($i:expr, $submac:ident!( $($args:tt)* )) => {
-        alt!($i,
-            parens!($submac!($($args)*)) => { |(_, content)| (Delimiter::Parenthesis, content) } |
-            braces!($submac!($($args)*)) => { |(_, content)| (Delimiter::Brace, content) } |
-            brackets!($submac!($($args)*)) => { |(_, content)| (Delimiter::Bracket, content) }
-        )
+    ($content:ident in $cursor:expr) => {
```
Comment (dtolnay) [inline src/parser.rs]: Looks like this code was copied out of syn -- all of these functions are internal unstable APIs and are subject to change at any time.

**Comment by lukaslueg** [comment]
`'a` should be parsed at one Literal as whitespace in between is not allowed. E.g. `<'a>` in the `macro_rules!()` can be matched by `< 'a >`, `<'a >`, `<    'a>` but not `<'  a>`.

**Comment by lukaslueg** [comment]
I've reviewed this and I'd like to go with @dtolnay 's PR, which seems more close to how `syn` was designed for.  @jmhain thanks very much for this PR, I know it's frustrating if a PR does not get pulled. I'd highly appreciate if you could tackle the remaining problems with the Parser, the multi-character Punct being the most pressing.

**Pull request closed without merging**